### PR TITLE
feat: Add 'open-namespaces' feature

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This is a step towards #13576

### CODE REVIEW

File: tests/testsuite/open_namespaces.rs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+                [package]
+                name = "foo::bar"
+                version = "0.0.1"
+                edition = "2015"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("read-manifest")
```
Comment (weihanglo) [inline tests/testsuite/open_namespaces.rs:53]: Why chose `cargo read-manifest` over `cargo metadata`?

**Review by weihanglo — APPROVED** [review]
Thanks. Just one question about `read-manifest`.

File: tests/testsuite/open_namespaces.rs (hunk 1 above)
Comment (epage) [reply tests/testsuite/open_namespaces.rs:53]: Less output to deal with.

`read-manfiest` is a short term command.  I expect to replace this with `check` as we work to get compilation working.

File: crates/cargo-util-schemas/src/restricted_names.rs [hunk 2]
```diff
@@ -84,6 +89,17 @@ pub(crate) fn validate_package_name(name: &str, what: &'static str) -> Result<()
 
 /// Ensure a package name is [valid][validate_package_name]
 pub(crate) fn sanitize_package_name(name: &str, placeholder: char) -> String {
+    let mut slug = String::new();
```
[the remark this answers is not in the document]
Comment (epage) [reply crates/cargo-util-schemas/src/restricted_names.rs:92]: `-Zscript` is also unstable :)

File: tests/testsuite/open_namespaces.rs [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+      "crate_types": [
+        "lib"
+      ],
+      "doc": true,
+      "doctest": true,
+      "edition": "2015",
+      "kind": [
+        "lib"
+      ],
+      "name": "foo::bar",
```
Comment (Nemo157) [inline tests/testsuite/open_namespaces.rs:90]: My understanding from the RFC was that the library name was meant to just be `bar`, that would avoid the rustc issue you noted

> The autogenerated `lib.name` key for such a crate will just be bar, the leaf crate name, and the expectation is that to use such crates one must use `--extern foo::bar=bar.rlib` syntax. There may be some better things possible here, perhaps `foo_bar` can be used here.
> —<https://rust-lang.github.io/rfcs/3243-packages-as-optional-namespaces.html#reference-level-explanation>

Comment (Nemo157) [reply tests/testsuite/open_namespaces.rs:90]: (I'd assume this also applies to the autogenerated `bin.name` for `src/main.rs`)

**Comment by tisonkun** [comment]
@epage @weihanglo Is there an easy way I enable this feature in a workspace? Now it seems I need to add `cargo-features = ["open-namespaces"]` to each crates in the workspace and I have more than 20 crates in the workspace ..

**Comment by weihanglo** [comment]
No.

Just note that at this moment this feature is not useful at all.

**Comment by epage** [comment]
To add, only enough has been implemented to unblock compiler work.

**Comment by tisonkun** [comment]
Thank you! Yes I see https://github.com/rust-lang/rust/issues/122349#issuecomment-2002208868 and hit exactly the same now.

**Pull request merged**